agent/stats/engine.go

```diff
@@ -438,16 +439,23 @@ func (engine *DockerStatsEngine) GetInstanceMetrics(includeServiceConnectStats b
 		}
 	}
 
-	for taskArn := range engine.tasksToContainers {
+	taskStatsToCollect := engine.getTaskStatsToCollect()
+	for taskArn := range taskStatsToCollect {
 		containerMetrics, err := engine.taskContainerMetricsUnsafe(taskArn)
 		if err != nil {
 			seelog.Debugf("Error getting container metrics for task: %s, err: %v", taskArn, err)
-			continue
+			// skip collecting service connect related metrics, if task is service connect enabled
+			if _, scTask := engine.taskToServiceConnectStats[taskArn]; !scTask {
+				continue
+			}
 		}
 
 		if len(containerMetrics) == 0 {
 			seelog.Debugf("Empty containerMetrics for task, ignoring, task: %s", taskArn)
-			continue
+			// skip collecting service connect related metrics, if task is service connect enabled
+			if _, scTask := engine.taskToServiceConnectStats[taskArn]; !scTask {
```

Comment:
Could we compute scTask just once at the beginning of the for loop and use that in the 2 if conditions

Reply:
This looks like it could be a data race, as some other process is going to be writing to this map at the same time you are reading it. Is there a better way to check if the task is not service connect enabled, or maybe wrap this into a helper function with a mutex?